src/txmempool.cpp

```diff
@@ -837,3 +869,45 @@ const CTxMemPool::setEntries & CTxMemPool::GetMemPoolChildren(txiter entry) cons
     assert(it != mapLinks.end());
     return it->second.children;
 }
+
+CFeeRate CTxMemPool::GetMinFee(size_t sizelimit) const {
+    LOCK(cs);
+    if (!blockSinceLastRollingFeeBump || rollingMinimumFeeRate == 0)
+        return CFeeRate(rollingMinimumFeeRate);
+
+    int64_t time = GetTime();
+    if (time > lastRollingFeeUpdate + 10) {
+        double halflife = ROLLING_FEE_HALFLIFE;
+        if (DynamicMemoryUsage() < sizelimit / 4)
+            halflife /= 4;
+        else if (DynamicMemoryUsage() < sizelimit / 2)
+            halflife /= 2;
+
+        rollingMinimumFeeRate = rollingMinimumFeeRate / pow(2.0, (time - lastRollingFeeUpdate) / halflife);
+        lastRollingFeeUpdate = time;
+
+        if (rollingMinimumFeeRate < ::minRelayTxFee.GetFeePerK())
```

Comment:
On further thought -- would it make more sense to just move this code out of the mempool and into main.cpp, to isolate the mempool from relay policy?  We could make `TrimToSize()` return the fee rate of the last package it removes, and then leave `AcceptToMemoryPool()` responsible for deciding what to do with the prevailing relay fee after eviction (including this logic for decaying things back down).